bbc/programmes-frontend
# [PROGRAMMES-6351] - Added UAS module to send SMP player activities

## UAS module Specs:
- [x] UAS service only send data to Activity server when the user is logged in (by default), but it includes some parameters to send data when the user is also logged out (configurable in code).
These requests look like this:

| Code response | Method | Type | Protocol | Url | Call |
|---|---|---|---|---|---|
| 202 | POST | plays | https | activity.test.api.bbc.co.uk | xhr |
| 202 | POST | heartbeat | https | heartbeat.activity.test.api.bbc.co.uk | xhr |
| 202 | POST | reads | https | activity.test.api.bbc.co.uk | xhr |

- [x] UAS send different activities: **started, paused, ended, reads, heartbeat**. Example:
```json
{
"action" : "started",
"actionContext" : "urn:bbc:tv:version_offset:p06f8f0f#22",
"resourceDomain" : "tv",
"resourceId" : "p06f8f0b",
"resourceType" : "clip",
} 
```
- [x] "Heartbeat" event is sent every 30 seconds
- [x] "Reads" event is sent when the user visit the page through a recommendation and is loggedin
- [x] UAS only runs in clip page.
- [x] UAS only send data for int/test/stage/live environments.
- [x] Uas stop sending data to Activity server for a few minutes if it receives error responses (A.K.A circuitbreaker in js)




## To test it:
1. Go to an INT v3 clip page like: https://programmes-frontend.int.api.bbc.co.uk/programmes/p06f8f0b?__flush_cache=1
2. Login
3. refresh the page after login. You should see a the next requests similar to the table above.


## Question:
Uas module use an ApiKey when sending activities to the server:
https://github.com/bbc/programmes-frontend/blob/uas-module/src/Ds2013/Presenters/Utilities/SMP/SmpPresenter.php#L132
Not sure if is useful to store it in any secure place as is gonna be dumped later in clear text in the javascript DOM. In V2 this is the case.


<details>
  <summary>Expand information about the implementation of UAS module:</summary>

I tried to find an existing UAS client that we could use. However I was no lucky. As a prove of this you can see some interesting options I found, but all of them were not reliable as they are done in scala, node, etc:
* https://github.com/bbc/rms-uas-client
* https://github.com/bbc/ibl-user-activity
* https://github.com/bbc/ibl-user-activity-client
* https://github.com/bbc/node-uas-client

From my conversations with some teams I got:
- IBL team doesn't use any UAS client, instead they send data to IBL, and IBL transform it into UAS format. However /programmes wouldn't be able to use IBL as it transform into an IPlayer specific way.
- Iplayer team use a backend UAS client only usable by Java or scala systems.

UAS client Confluence page says that we can just require it on the fly, this might be a reliable solution), however I didn't do it as I assume there is a reason to not be required in V2:
https://confluence.dev.bbc.co.uk/display/activity/User+Activity+Client+Library+User+Guide#UserActivityClientLibraryUserGuide-Gettingstarted

I finished choosing the existing V2:  https://github.com/bbc/uas-client-pal
</details>

<details>
  <summary>How I review this?</summary>
  In order to make a review you must know that the only file I could refactor (clean and improve properly) was UasService.js
This file acts as a façade to UAS module and as you can imagine this UasService relies in the UasClient to send activity user.
</details>
